database.cc

```diff
@@ -2321,6 +2321,64 @@ future<> database::drain() {
     co_await _commitlog->shutdown();
 }
 
+future<std::unordered_map<sstring, std::unordered_map<std::pair<sstring, sstring>, db::snapshot_ctl::snapshot_details, utils::tuple_hash>>> database::get_snapshot_details() {
+    std::vector<sstring> data_dirs = _cfg.data_file_directories();
+    auto dirs_only_entries = lister::dir_entry_types{directory_entry_type::directory};
+    std::unordered_map<sstring, std::unordered_map<std::pair<sstring, sstring>, snapshot_ctl::snapshot_details, utils::tuple_hash>> details;
+
+    for (auto& datadir : data_dirs) {
+        co_await lister::scan_dir(datadir, dirs_only_entries, [this, &dirs_only_entries, &details] (fs::path parent_dir, directory_entry de) mutable -> future<> {
+            // KS directory
+            sstring ks_name = de.name;
+
+            co_return co_await lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) mutable -> future<> {
```

Comment:
When you're using `co_await`, everything in the coroutine's frame is kept alive. This distinguishes `co_return co_await e` from `return e` (where `e` is an expression giving a future): when you `return`, the function's frame is destroyed, but when you `co_return co_await`, you first wait for `e` to finish (i.e. the future must resolve), and only then the frame is destroyed. So for `co_return co_await`, you can depend on local variables to remain valid, unlike when you `return`.

On the other hand `co_await` turns your function into a coroutine, causing an additional heap allocation (coroutine frame).

In this case you would probably be fine with a simple `return` since you don't need to keep anything alive.
However this is not performance critical code so I guess `co_return co_await` is also OK, and it makes the code more consistent. So depends on your preference.

Reply:
Same comment applies to the `co_return co_await` expressions below - you don't capture references to local variables of your coroutines, so they could as well be functions.

Reply:
I left it as is for consistency